### PR TITLE
feat: support GLM-5.2

### PR DESCRIPTION
**Pull request by CUHKSZzxy**
## Motivation

Add GLM-5.2 support to the PyTorch backend, including DSA, MTP, routed-expert replay, and serving integration. The shared DSA changes also improve DeepSeek-V3.2 correctness and decode efficiency.

## Changes

- Add GLM-5.2 config/model loading, interleaved RoPE, MoE-only FP8, and tool calls emitted directly from reasoning.
- Implement DSA indexing for GLM-5.2 and DeepSeek-V3.2: radix top-k, fused indexer preparation, shared top-k reuse, and BF16/FP8 sparse FlashMLA cache paths. `LMDEPLOY_DISABLE_DSA_INDEXER_FUSION=1` keeps the original path.
- Add GLM-5.2 MTP with aligned draft distribution, shared embeddings, post-norm hidden-state flow, index sharing, and multi-token query handling.
- Add routed-expert capture/replay through prefix caching and OpenAI/Anthropic serving APIs.
- Centralize Transformers config fallback and add coverage for model config, CUDA kernels, cache/attention, MTP/routed-expert replay, prefix caching, and response parsing.

### Academic

| Dataset | Metric | LMDeploy | vLLM | LMDeploy − vLLM |
|---|---|---:|---:|---:|
| MMLU-Pro | Accuracy | 86.86% | 87.01% | -0.15 pp |
| AIME 2025 | Accuracy, 32-run average | 93.33% | 93.54% | -0.21 pp |
| GPQA Diamond | Accuracy, 4-run average | 90.03% | 90.03% | 0.00 pp |

with MTP 5

| Dataset | Metric | LMDeploy |
|---|---|---:|
| AIME 2025 | Accuracy, 32-run average | 93.54% |
| GPQA Diamond | Accuracy, 4-run average | 92.42% |

Both engines were evaluated with TP8 under identical settings on AutoToolChain.

Generation used `temperature=1.0`, `top_p=0.95`, and a maximum generation length of `163,840` tokens, following the [GLM-5.2 model card](https://huggingface.co/zai-org/GLM-5.2). Scores may differ from the published results because other evaluation details noted in the model card, such as the reasoning-task system prompt, may not have been fully aligned.

### GSM8k

| Engine / topology | MTP | Correct / total | Accuracy | 95% Wilson CI | Errors | Empty | Cap hits |
|---|---:|---:|---:|---:|---:|---:|---:|
| LMDeploy TP8 | off | 1266 / 1319 | 95.98% | 94.78–96.91% | 0 | 0 | 2 |
| LMDeploy TP8 | 5 tokens | 1267 / 1319 | 96.06% | 94.87–96.98% | 0 | 0 | 1 |
| LMDeploy DP8+EP8 | off | 1272 / 1319 | 96.44% | 95.29–97.31% | 0 | 0 | 1 |
| LMDeploy DP8+EP8 | 5 tokens | 1271 / 1319 | 96.36% | 95.21–97.24% | 0 | 0 | 1 |
| vLLM TP8 | off | 1271 / 1319 | 96.36% | 95.21–97.24% | 0 | 0 | 0 |
| vLLM TP8 | 5 tokens | 1265 / 1319 | 95.91% | 94.70–96.85% | 0 | 0 | 3 |

### Aggregate results

Short aggregate per-position acceptance rates:

| Engine | Position 0 | Position 1 | Position 2 | Position 3 | Position 4 |
|---|---:|---:|---:|---:|---:|
| LMDeploy | 77.95% | 49.62% | 32.32% | 19.58% | 10.46% |
| vLLM | 75.05% | 48.42% | 30.35% | 18.81% | 11.92% |

Long per-position acceptance rates:

| Workload | Engine | Position 0 | Position 1 | Position 2 | Position 3 | Position 4 |
|---|---|---:|---:|---:|---:|---:|
| >4K | LMDeploy | 73.50% | 41.55% | 22.89% | 12.91% | 4.59% |
| >4K | vLLM | 72.16% | 46.57% | 25.29% | 12.73% | 5.64% |
| >8K | LMDeploy | 77.10% | 48.93% | 27.53% | 14.50% | 7.39% |
| >8K | vLLM | 74.12% | 45.18% | 23.91% | 11.81% | 5.08% |
| >16K | LMDeploy | 76.70% | 48.01% | 26.74% | 14.92% | 8.13% |
| >16K | vLLM | 74.02% | 45.07% | 25.14% | 13.39% | 7.15% |

### Throughput

| Output length | Requests | TP8 output tok/s | DP8+EP8 output tok/s | Speedup | TP8 duration | DP8+EP8 duration |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| Natural | 10,000 | 3,228.23 | 4,831.38 | **1.50x** | 610.01 s | 407.60 s |
| 2,048 | 8,000 | 2,582.32 | 6,036.53 | **2.34x** | 6,344.68 s | 2,714.14 s |
| 4,096 | 8,000 | 2,145.58 | 4,049.21 | **1.89x** | 15,272.31 s | 8,092.45 s |
| 8,192 | 4,000 | 1,594.37 | 3,035.30 | **1.90x** | 20,552.32 s | 10,795.63 s |
| 16,384 | 1,000 | 1,012.60 | 1,605.38 | **1.59x** | 16,180.10 s | 10,205.66 s |
| 32,768 | 500 | 629.18 | 991.59 | **1.58x** | 26,040.29 s | 16,522.91 s |

### Latency

| Output length | TP8 mean TTFT | DP8+EP8 mean TTFT | TP8 mean TPOT | DP8+EP8 mean TPOT |
| ---: | ---: | ---: | ---: | ---: |
| Natural | 247.93 s | 137.65 s | 306.85 ms | 399.75 ms |
| 2,048 | 2,832.64 s | 1,022.85 s | 253.28 ms | 280.64 ms |
| 4,096 | 7,038.16 s | 3,117.33 s | 194.88 ms | 315.44 ms |
| 8,192 | 7,136.28 s | 2,109.02 s | 590.41 ms | 542.26 ms |
| 16,384 | 12.34 s | 7.43 s | 706.87 ms | 452.57 ms |
| 32,768 | 6.54 s | 5.17 s | 660.96 ms | 413.83 ms |

DP8+EP8 config: max batch size 256, cache 0.6

### Five-layer ablations

Matched five-layer GLM-5.2 FP8 ablations on the same H200 pair: TP2, BF16 MLA KV cache, CUDA graphs enabled, and MTP disabled. Throughput is the mean of three post-warmup batch-256 runs with 128 input and 2,048 forced output tokens (CV ≤ 0.73%). Timeline measurements average both ranks from a one-second batch-128 steady-decode capture.

| Optimization | Median cycle | Cycle change | CUDA nodes/cycle | Node change | Output throughput | Gain |
|---|---:|---:|---:|---:|---:|---:|
| Baseline | 16.634 ms | — | 268 | — | 14,240 tok/s | — |
| Zero-copy BF16 sparse MLA | 4.512 ms | -72.87% | 258 | -3.73% | 42,122 tok/s | +195.79% |
| Compact blocked-FP8 MoE scheduling | 16.457 ms | -1.06% | 280 | +4.48% | 14,328 tok/s | +0.62% |

Both optimizations are measured independently against the same baseline.

The previous BF16 sparse path materialized the non-contiguous paged main-K cache before attention. In a representative five-layer draft forward, the whole-cache copy took 1.876 ms of the 3.027 ms forward; the matched FP8 path had no corresponding copy and took 1.394 ms. The BF16 sparse-attention kernel itself was faster—the avoidable cache materialization was the bottleneck addressed by the zero-copy strided view.

### DSA indexer fusion

Full-model GLM-5.2 FP8 TP8 timeline comparison:

| Metric | Fusion off | Fusion on | Change |
|---|---:|---:|---:|
| Forward duration | 27.345 ms | 26.585 ms | -2.78% |
| CUDA-graph child kernels | 4,455 | 4,182 | -273 / -6.13% |

The full model has 21 indexer layers, so fused Q/K preprocessing removes exactly 13 CUDA-graph child kernels per layer (`21 × 13 = 273`).

Same idea as **Indexer Prologue Fusion** in [SGLang blog](https://www.lmsys.org/blog/2026-07-13-glm52-optimization/)

## Validation

- Pre-commit hooks and targeted replay, prefix-cache, engine, and serving tests pass locally (218 tests).
- Real 5-layer FP8 smoke tests passed for GLM-5.2 and DeepSeek-V3.2.
- Indexer fusion improved GLM-5.2 5-layer MTP batch-256 median output throughput by 1.12% across three profiler-free runs; short batch-8 timelines reduced median cycle time by 3.41% for GLM-5.2 and 4.66% for DeepSeek-V3.2.

## Assistance

Assisted with Codex + GPT-5.6-Sol xHigh, reviewed manually

### CODE REVIEW

File: lmdeploy/pytorch/backends/cuda/graph_runner.py [hunk 1]
```diff
@@ -240,7 +240,14 @@ def get_graph_key(self, input_ids: torch.Tensor, position_ids: torch.Tensor, pas
             batch_size = self._get_capture_tokens(batch_size)
         else:
             batch_size = self._get_capture_tokens(meta.padding_batch_size)
-        return (batch_size, is_decoding, enable_microbatch, query_len)
+        graph_key = (batch_size, is_decoding, enable_microbatch, query_len)
+        if 'skip_topk' in kwargs:
```
Comment (grimoire) [inline lmdeploy/pytorch/backends/cuda/graph_runner.py]: I think we can create a function call in models/utils/cudagraph.py and override it in modeling to add these "model sepcialized" key.

File: lmdeploy/pytorch/backends/cuda/nsa.py [hunk 2]
```diff
@@ -20,18 +55,42 @@ def __init__(self, topk: int, softmax_scale: float, block_size: int, fill: int)
...
+                           cu_seqlen_q,
+                           k_seqlens,
+                           block_offset,
+                           max_q_seqlen=max_q_seqlen,
+                           max_k_seqlen=max_kv_seqlen,
+                           causal=True)
+        # Multi-token queries need one causal KV length per score row.
+        if scores.size(0) != k_seqlens.size(0):
+            k_seqlens = _get_causal_k_seqlens(cu_seqlen_q, q_seqlens, k_seqlens, scores.size(0))
+        sparse_index_topk = _get_sparse_index_topk(self.topk)
```
Comment (grimoire) [inline lmdeploy/pytorch/backends/cuda/nsa.py]: why don't we get the function in init?

File: lmdeploy/pytorch/backends/cuda/nsa.py [hunk 3]
```diff
@@ -20,18 +55,42 @@ def __init__(self, topk: int, softmax_scale: float, block_size: int, fill: int)
...
+                           k_s_cache[..., 0],
+                           cu_seqlen_q,
+                           k_seqlens,
+                           block_offset,
+                           max_q_seqlen=max_q_seqlen,
+                           max_k_seqlen=max_kv_seqlen,
+                           causal=True)
+        # Multi-token queries need one causal KV length per score row.
+        if scores.size(0) != k_seqlens.size(0):
+            k_seqlens = _get_causal_k_seqlens(cu_seqlen_q, q_seqlens, k_seqlens, scores.size(0))
```
Comment (grimoire) [inline lmdeploy/pytorch/backends/cuda/nsa.py]: This value won't change in single forward, right? should we place it in meta so we only need to compute it once?

File: lmdeploy/pytorch/kernels/cuda/dsa_indexer_preprocess.py [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+    Weights,
+    Cos,
+    Sin,
+    QOut,
+    QScaleOut,
+    num_heads: tl.constexpr,
+    score_scale: tl.constexpr,
+    fp8_min: tl.constexpr,
+    fp8_max: tl.constexpr,
+    stride_qt: tl.constexpr,
```
Comment (grimoire) [inline lmdeploy/pytorch/kernels/cuda/dsa_indexer_preprocess.py:50]: I have not run the real model yet. Would the stride change for different token numbers? tl.contexpr for frequent change value might leads to repeated compilation.

File: lmdeploy/pytorch/model_inputs.py [hunk 5]
```diff
@@ -406,7 +406,12 @@ def new(
 
         # position ids
         attention_mask, position_ids = cls.get_mask_and_position_ids(inputs)
-        q_start_loc = q_seqlens.cumsum(0) - q_seqlens
+        if inputs.max_q_seqlen == 1:
```
Comment (grimoire) [inline lmdeploy/pytorch/model_inputs.py]: This is true, but this won't bring much benefits, right?

**Review by grimoire — APPROVED** [review]
LGTM

File: docker/prepare_3rdparty_wheel.sh [hunk 6]
```diff
@@ -33,6 +34,10 @@ pip wheel -v --no-build-isolation --no-deps -w /wheels "git+https://github.com/d
 # sm100 compilation for Flash MLA requires NVCC 12.9 or higher
 FLASH_MLA_DISABLE_SM100=1 pip wheel -v --no-build-isolation --no-deps -w /wheels "git+https://github.com/deepseek-ai/FlashMLA.git@${FLASH_MLA_VERSION}"
 
+# fast_hadamard_transform
+pip wheel -v --no-build-isolation --no-deps -w /wheels \
+    "git+https://github.com/Dao-AILab/fast-hadamard-transform.git@${FAST_HADAMARD_TRANSFORM_VERSION}"
```
Comment (lvhan028) [inline docker/prepare_3rdparty_wheel.sh:39]: Does this dependency require a specific CUDA version? Can cuda 13+ work?

Comment (CUHKSZzxy) [reply docker/prepare_3rdparty_wheel.sh:39]: Not sure, haven't tested with cu13. cu128 should work

I have already found one way to eliminate this dependency and have made it the default option for GLM5.2. But since DeepSeekV3.2, GLM5.2, DeepSeekV4 may still use it, so adding it here for convenience.

**Comment by lvhan028** [comment]
the test_docker issue will be handled in another PR

File: lmdeploy/pytorch/spec_decode/base.py [hunk 7]
```diff
@@ -18,10 +18,10 @@ def _build_draft_dist_ctx(dist_ctx: DistContext, specdecode_config: SpecDecodeCo
     if specdecode_config is None:
         return None
 
-    if specdecode_config.method == 'qwen3_5_mtp':
+    draft_dist_config = specdecode_config.dist_config
+    if draft_dist_config == dist_ctx.dist_config:
```
Comment (lvhan028) [inline lmdeploy/pytorch/spec_decode/base.py:22]: cc @RunningLeon
